### PR TITLE
Handle cmd2

### PR DESCRIPTION
**Pull request by AdelaCorbeanu**

### CODE REVIEW

File: src/daemon/daemon.c [hunk 1]
```diff
@@ -18,23 +25,88 @@
...
+		current->parent = malloc(sizeof(*current->parent));
+		current->parent = last[job_count];
+		last[job_count]->number_subdir++;
+		last[job_count]->bytes += sb->st_size;
+		current->number_subdir = 0;
+		current->path = malloc(strlen(fpath) * sizeof(char) + 1);
+		strcpy(current->path, fpath); 
+		current->bytes = 0;
+		current->number_files = 0;	
+		last[job_count]->children[last[job_count]->number_subdir] = current;
```
Comment (Qmpzlawasd) [inline src/daemon/daemon.c]: `last[job_count]->number_subdir` starts from 1?

File: src/daemon/daemon.c [hunk 2]
```diff
@@ -99,7 +170,38 @@ int get_job_info(int64_t id) {}
...
+
+	if (ret) {
+		perror("Error creating thread");
+		exit(EXIT_FAILURE);
+	}
+
+	pthread_join(jobs[job_count]->thread, NULL);
+
+	jobs[job_count]->root = malloc(sizeof(*jobs[job_count]->root));
+	jobs[job_count]->root = last[job_count];
```
Comment (Qmpzlawasd) [inline src/daemon/daemon.c]: useless malloc, it's just copying a pointer.

**Review by Qmpzlawasd — CHANGES_REQUESTED** [review]

File: src/common/ipc.h [hunk 3]
```diff
@@ -7,13 +7,13 @@
...
-#define CMD_SUSPEND 2
+#define CMD_SUSPEND 1
+#define CMD_REMOVE  2
 #define CMD_RESUME  3
-#define CMD_REMOVE  4
+#define CMD_PRINT   4
 #define CMD_INFO    5
 #define CMD_LIST    6
-#define CMD_PRINT   7
+#define CMD_ADD     7
```
Comment (tezkerek) [inline src/common/ipc.h]: Revert

File: src/daemon/daemon.c [hunk 4]
```diff
@@ -18,23 +25,86 @@
...
+
+struct Job {
     pthread_t thread;
-    int8_t status; // status of job -> in progress(0), done(1), removed(3), paused(2)
+    int8_t status;           // status of job -> in progress(0), done(1), removed(3), paused(2)
     struct Directory *root; // children directories
 };
 
 static uint64_t job_count = 0;
 struct Job job_history[MAX_THREADS];
```
Comment (tezkerek) [inline src/daemon/daemon.c]: `job_history` is unused?

File: src/daemon/daemon.c [hunk 5]
```diff
@@ -18,23 +25,86 @@
...
 
+struct Job *jobs[MAX_THREADS];
+
+struct Directory *last[MAX_THREADS];
+
+static int build_tree (const char* fpath, const struct stat *sb, int typeflag, struct FTW *ftwbuf) {
+
+	if (!strcmp(fpath, last[job_count]->path)) return 0;
+
+	char *aux = malloc(strlen(fpath) * sizeof(char) + 1);
```
Comment (tezkerek) [inline src/daemon/daemon.c]: `(strlen(fpath) + 1) * sizeof(*current->path)`
Same for the other 2.

File: src/daemon/daemon.c [hunk 6]
```diff
@@ -18,23 +25,86 @@
...
+		last[job_count]->number_files++;
+		last[job_count]->bytes += sb->st_size;
+	}
+
+	return 0;
+}
+
+void *traverse (void *path) {
+
+	int nopenfd = 20;  // ne gandim cat vrem sa punem
```
Comment (tezkerek) [inline src/daemon/daemon.c]: `static`

File: src/daemon/daemon.c [hunk 7]
```diff
@@ -99,7 +169,37 @@ int get_job_info(int64_t id) {}
...
+	struct sched_param param;
+
+	int ret = pthread_attr_init (&tattr);
+	ret = pthread_attr_getschedparam (&tattr, &param);
+
+	param.sched_priority = priority;
+
+	ret = pthread_attr_setschedparam (&tattr, &param);
+
+	jobs[job_count] = malloc(sizeof(jobs[job_count]));
```
Comment (tezkerek) [inline src/daemon/daemon.c]: Shouldn't this be `sizeof(*jobs[job_count])`?

File: src/client/client.c [hunk 8]
```diff
@@ -36,7 +36,7 @@ int send_ipc_msg(int serverfd, int8_t cmd, char *payload, int64_t payload_len) {
 
 int main(int argc, char *argv[]) {
     int serverfd = connect_to_socket();
-
+
```
Comment (tezkerek) [inline src/client/client.c]: Revert

File: src/daemon/daemon.c [hunk 9]
```diff
@@ -18,23 +25,86 @@
...
+	struct Directory *root = malloc(sizeof(*root));
+	root->path = malloc(strlen(p) * sizeof(char) + 1);
+	strcpy(root->path, p);
+	root->parent = NULL;
+	root->bytes = root->number_files = root->number_subdir = 0;
+	last[job_count] = root;
+
+	if (nftw(p, build_tree, nopenfd, FTW_PHYS) == -1) {
+		perror("nsfw");   // modif
+		return errno;
```
Comment (tezkerek) [inline src/daemon/daemon.c]: Returning `int` in a `void*` function.

**Review by tezkerek — CHANGES_REQUESTED** [review]
Run `clang-format` or VS Code formatter.

File: src/daemon/daemon.c [hunk 10]
```diff
@@ -18,23 +25,86 @@
...
 static uint64_t job_count = 0;
 struct Job job_history[MAX_THREADS];
 
+struct Job *jobs[MAX_THREADS];
+
+struct Directory *last[MAX_THREADS];
+
+static int build_tree (const char* fpath, const struct stat *sb, int typeflag, struct FTW *ftwbuf) {
+
+	if (!strcmp(fpath, last[job_count]->path)) return 0;
```
Comment (tezkerek) [inline src/daemon/daemon.c]: Why `job_count`? That always refers to the last job.

File: src/daemon/daemon.c [hunk 11]
```diff
@@ -18,23 +25,86 @@
...
-{
+
+struct Job {
     pthread_t thread;
-    int8_t status; // status of job -> in progress(0), done(1), removed(3), paused(2)
+    int8_t status;           // status of job -> in progress(0), done(1), removed(3), paused(2)
     struct Directory *root; // children directories
 };
 
 static uint64_t job_count = 0;
```
Comment (tezkerek) [inline src/daemon/daemon.c:16]: This should be made thread-safe. Let's use [atomics](https://en.cppreference.com/w/c/atomic).

File: src/daemon/daemon.c [hunk 12]
```diff
@@ -18,23 +25,86 @@
...
 struct Job job_history[MAX_THREADS];
 
+struct Job *jobs[MAX_THREADS];
```
Comment (tezkerek) [inline src/daemon/daemon.c:18]: The daemon could receive two simultaneous commands for the same job. We'll probably need a lock for each job.

File: src/daemon/daemon.c [hunk 13]
```diff
@@ -99,7 +169,37 @@ int get_job_info(int64_t id) {}
...
+
+	ret = pthread_attr_setschedparam (&tattr, &param);
+
+	jobs[job_count] = malloc(sizeof(jobs[job_count]));
+	jobs[job_count]->status = JOB_STATUS_IN_PROGRESS;
+	ret = pthread_create (&jobs[job_count]->thread, NULL, traverse, (void*)path);
+
+	if (ret) {
+		perror("Error creating thread");
+		exit(EXIT_FAILURE);
```
Comment (tezkerek) [inline src/daemon/daemon.c]: Don't exit, return -1.

**Review by tezkerek — CHANGES_REQUESTED** [review]

File: src/daemon/thread_utils.c [hunk 14]
```diff
@@ -25,62 +25,96 @@ void check_suspend(struct Job *job_to_check) {
...
     }
 
-    return 0;
+    return new_directory;
 }
 
-void *traverse(void *path) {
+void *traverse(void *args) {
+    char *path = ((struct traverse_args *)args)->path;
+    int job_id = ((struct traverse_args *)args)->job_id;
```
Comment (tezkerek) [inline src/daemon/thread_utils.c]: `int64_t job_id`.
We should probably add a typedef for this.

File: src/daemon/thread_utils.c [hunk 15]
```diff
@@ -25,62 +25,96 @@ void check_suspend(struct Job *job_to_check) {
...
-    root->bytes = root->number_files = root->number_subdir = 0;
-    last[job_count] = root;
+    FTS *ftsp;
+    FTSENT *p, *chp;
 
-    if (nftw(p, build_tree, nopenfd, FTW_PHYS) == -1) {
-        perror("nsfw"); // modif
+    char **fts_args = malloc(2 * sizeof(char *));
+    fts_args[0] = path;
+    fts_args[1] = NULL;
```
Comment (tezkerek) [inline src/daemon/thread_utils.c:91]: `char *fts_args[] = {path, NULL}`

File: src/daemon/thread_utils.c [hunk 16]
```diff
@@ -25,62 +25,96 @@ void check_suspend(struct Job *job_to_check) {
...
-                      struct FTW *ftwbuf) {
+/**
+ * Creates a new directory
+ * path - the path of the new directory
+ * parent - the parent of the new directory, 
+ * should be NULL if the new directory is root
+ */
+struct Directory *create_directory(char *path, struct Directory *parent) {
+    struct Directory *new_directory = malloc(sizeof(struct Directory));
+    new_directory->path = malloc((strlen(path) + 1) * sizeof(char));
```
Comment (tezkerek) [inline src/daemon/thread_utils.c:52]: `sizeof(*new_directory->path)`
Also use `da_malloc` everywhere.

File: src/daemon/daemon.c [hunk 17]
```diff
@@ -94,23 +94,24 @@ int create_job(char *path, int8_t priority) {
 
     ret = pthread_attr_setschedparam(&tattr, &param);
 
-    jobs[job_count] = malloc(sizeof(jobs[job_count]));
-    jobs[job_count]->status = JOB_STATUS_IN_PROGRESS;
-    ret = pthread_create(&jobs[job_count]->thread, NULL, traverse, (void *)path);
+    // TODO: Mutex lock?
+    int job_id = job_count;
+    ++job_count;
+    // TODO: Mutex unlock?
```
Comment (tezkerek) [inline src/daemon/daemon.c:66]: Regarding locking, atomics seem pretty easy. In this situation you need https://en.cppreference.com/w/c/atomic/atomic_fetch_add.

File: src/daemon/daemon.c [hunk 18]
```diff
@@ -94,23 +94,24 @@ int create_job(char *path, int8_t priority) {
...
+
+    struct traverse_args *args = malloc(sizeof(struct traverse_args));
```
Comment (tezkerek) [inline src/daemon/daemon.c:68]: `sizeof(*args)`

File: src/daemon/thread_utils.c [hunk 19]
```diff
@@ -25,62 +25,96 @@ void check_suspend(struct Job *job_to_check) {
...
-                      int typeflag,
-                      struct FTW *ftwbuf) {
+/**
+ * Creates a new directory
+ * path - the path of the new directory
+ * parent - the parent of the new directory, 
+ * should be NULL if the new directory is root
+ */
+struct Directory *create_directory(char *path, struct Directory *parent) {
+    struct Directory *new_directory = malloc(sizeof(struct Directory));
```
Comment (tezkerek) [inline src/daemon/thread_utils.c:51]: `sizeof(*new_directory)`

**Review by tezkerek — CHANGES_REQUESTED** [review]

File: src/daemon/thread_utils.c [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+
+    if (parent != NULL) {
+      // Add the new directory to the list of his parents children
+      if (parent->children_capacity == parent->number_subdir) {
+        if (parent->children_capacity == 0)
+            parent->children_capacity = 1;
+        else
+            parent->children_capacity *= 2;
+
+        parent->children = realloc(parent->children, parent->children_capacity);
```
Comment (Qmpzlawasd) [inline src/daemon/thread_utils.c:51]: Manpage:
> void *realloc(void *ptr, size_t size);  
> The  realloc() function changes the size of the memory block pointed by ptr to size bytes.

Should be `realloc(parent->children, parent->children_capacity * sizeof(struct Directory *));`

**Review by Qmpzlawasd — CHANGES_REQUESTED** [review]

File: src/daemon/thread_utils.c (hunk 20 above)
Comment (Qmpzlawasd) [inline src/daemon/thread_utils.c]: Should be `realloc(parent->children, parent->children_capacity * sizeof(struct Directory *));`
manpage:
>  `If new size is larger than the old size, the added memory will not be initialized.`

**Review by Qmpzlawasd — CHANGES_REQUESTED** [review]

**Comment by tezkerek** [comment]
Continued in #19.

**Pull request closed without merging**